### ISSUE TITLE
Redis with Sentinel unable to failover with error Failed to resolve hostname

### ISSUE THREAD

**Issue by SureshGota** [bitnami/containers#53232]
### Name and Version

bitnami/redis-sentinel:latest

### What architecture are you using?

None

### What steps will reproduce the bug?

Trying to run 3 redis servers(master-2 slaves) with 3 sentinels. When the master is down the sentinels are supposed to elect one of the slaves as master but instead, the sentinels complain that they are unable to resolve the hostname (**Failed to resolve hostname <prev master name>**)

```yaml
version: '3.9'

networks:
  redis-cluster-compose:
    driver: bridge


services:

  redis-master:
    container_name: redis-master
    image: bitnami/redis:latest
    environment:
      - REDIS_REPLICATION_MODE=master
      - REDIS_PASSWORD=redispassword
    networks:
      - redis-cluster-compose
    volumes:
      - "./conf:/data"
    ports:
      - "6379:6379"

  redis-slave1:
    image: bitnami/redis:latest
    container_name: redis-slave1
    environment:
      - REDIS_REPLICATION_MODE=slave
      - REDIS_MASTER_HOST=redis-master
      - REDIS_MASTER_PASSWORD=redispassword
      - REDIS_PASSWORD=redispassword
    networks:
      - redis-cluster-compose
    command: redis-server --slaveof redis-master 6379
    ports:
      - "6380:6379"
    links:
      - redis-master
    volumes:
      - "./conf:/data"
      
      
  redis-slave2:
    image: bitnami/redis:latest
    container_name: redis-slave2
    environment:
      - REDIS_REPLICATION_MODE=slave
      - REDIS_MASTER_HOST=redis-master
      - REDIS_MASTER_PASSWORD=redispassword
      - REDIS_PASSWORD=redispassword
    networks:
      - redis-cluster-compose
    command: redis-server --slaveof redis-master 6379
    ports:
      - "6381:6379"
    links:
      - redis-master
    volumes:
      - "./conf:/data"
      
  redis-sentinel-1:
    image: 'bitnami/redis-sentinel:latest'
    container_name: sentinel-1
    networks:
      - redis-cluster-compose
    environment:
      - REDIS_MASTER_SET=redis-master
      - REDIS_MASTER_HOST=redis-master
      - REDIS_SENTINEL_DOWN_AFTER_MILLISECONDS=10000
      - REDIS_SENTINEL_RESOLVE_HOSTNAMES=yes
      - REDIS_MASTER_PASSWORD=redispassword
    depends_on:
      - redis-master
      - redis-slave1
      - redis-slave2
    ports: 
       - "26379:26379"
       
  redis-sentinel-2:
    image: 'bitnami/redis-sentinel:latest'
    container_name: sentinel-2
    networks:
      - redis-cluster-compose
    environment:
      - REDIS_MASTER_SET=redis-master
      - REDIS_MASTER_HOST=redis-master
      - REDIS_SENTINEL_DOWN_AFTER_MILLISECONDS=10000
      - REDIS_SENTINEL_RESOLVE_HOSTNAMES=yes
      - REDIS_MASTER_PASSWORD=redispassword
    depends_on:
      - redis-master
      - redis-slave1
      - redis-slave2
    ports: 
      - "26380:26379"
      
  redis-sentinel-3:
    image: ' bitnami/redis-sentinel:latest '
    container_name: sentinel-3
    networks:
      - redis-cluster-compose
    environment:
      - REDIS_MASTER_SET=redis-master
      - REDIS_MASTER_HOST=redis-master
      - REDIS_SENTINEL_DOWN_AFTER_MILLISECONDS=10000
      - REDIS_SENTINEL_RESOLVE_HOSTNAMES=yes
      - REDIS_MASTER_PASSWORD=redispassword
    depends_on:
      - redis-master
      - redis-slave1
      - redis-slave2
    ports: 
      - "26381:26379"
```

### What is the expected behavior?

Expected to have a successful failover and a slave promoted to the next master.

### What do you see instead?
**Sentinel 1**
```
2023-11-27 18:01:22 1:X 27 Nov 2023 12:31:22.758 # +vote-for-leader 048facec9b1df0577b62d4426289e96066078028 1
2023-11-27 18:01:45 1:X 27 Nov 2023 12:31:45.512 # Failed to resolve hostname 'redis-master'
2023-11-27 18:01:45 1:X 27 Nov 2023 12:31:45.571 # +tilt #tilt mode entered
2023-11-27 18:01:56 1:X 27 Nov 2023 12:31:56.705 # Failed to resolve hostname 'redis-master'
2023-11-27 18:01:56 1:X 27 Nov 2023 12:31:56.797 # +tilt #tilt mode entered
2023-11-27 18:02:07 1:X 27 Nov 2023 12:32:07.868 # Failed to resolve hostname 'redis-master'
2023-11-27 18:02:07 1:X 27 Nov 2023 12:32:07.953 # +tilt #tilt mode entered
```

**Sentinel 2**

```
2023-11-27 17:56:36 1:X 27 Nov 2023 12:26:36.059 * +sentinel sentinel 532f657eb78380bfe9786115c86bad91b870810e 172.23.0.6 26379 @ redis-master 172.23.0.2 6379
2023-11-27 17:56:36 1:X 27 Nov 2023 12:26:36.077 * Sentinel new configuration saved on disk
2023-11-27 18:00:44 1:X 27 Nov 2023 12:30:44.475 # Failed to resolve hostname 'redis-master'
2023-11-27 18:00:44 1:X 27 Nov 2023 12:30:44.553 # +tilt #tilt mode entered
2023-11-27 18:01:14 1:X 27 Nov 2023 12:31:14.604 # -tilt #tilt mode exited
2023-11-27 18:01:14 1:X 27 Nov 2023 12:31:14.604 # +sdown master redis-master 172.23.0.2 6379
2023-11-27 18:01:22 1:X 27 Nov 2023 12:31:22.682 # Failed to resolve hostname 'redis-master'
2023-11-27 18:01:22 1:X 27 Nov 2023 12:31:22.683 # +odown master redis-master 172.23.0.2 6379 #quorum 2/2
2023-11-27 18:01:22 1:X 27 Nov 2023 12:31:22.683 # +new-epoch 1
2023-11-27 18:01:22 1:X 27 Nov 2023 12:31:22.684 # +try-failover master redis-master 172.23.0.2 6379
2023-11-27 18:01:22 1:X 27 Nov 2023 12:31:22.710 * Sentinel new configuration saved on disk
2023-11-27 18:01:22 1:X 27 Nov 2023 12:31:22.710 # +vote-for-leader 048facec9b1df0577b62d4426289e96066078028 1
2023-11-27 18:01:22 1:X 27 Nov 2023 12:31:22.754 * 532f657eb78380bfe9786115c86bad91b870810e voted for 048facec9b1df0577b62d4426289e96066078028 1
2023-11-27 18:01:22 1:X 27 Nov 2023 12:31:22.758 * bdf80cedf0b9c746b96eca87c49431088d09d1f6 voted for 048facec9b1df0577b62d4426289e96066078028 1
2023-11-27 18:01:22 1:X 27 Nov 2023 12:31:22.795 # +tilt #tilt mode entered
2023-11-27 18:01:47 1:X 27 Nov 2023 12:31:47.384 # Failed to resolve hostname 'redis-master'
2023-11-27 18:01:47 1:X 27 Nov 2023 12:31:47.461 # +tilt #tilt mode entered
2023-11-27 18:01:54 1:X 27 Nov 2023 12:31:54.513 # Failed to resolve hostname 'redis-master'
2023-11-27 18:01:54 1:X 27 Nov 2023 12:31:54.572 # +tilt #tilt mode entered
2023-11-27 18:02:05 1:X 27 Nov 2023 12:32:05.639 # Failed to resolve hostname 'redis-master'
2023-11-27 18:02:05 1:X 27 Nov 2023 12:32:05.723 # +tilt #tilt mode entered
2023-11-27 18:02:16 1:X 27 Nov 2023 12:32:16.804 # Failed to resolve hostname 'redis-master'
2023-11-27 18:02:16 1:X 27 Nov 2023 12:32:16.871 # +tilt #tilt mode entered
```

**Comment by carrodher** [comment]
The issue may not be directly related to the Bitnami container image or Helm chart, but rather to how the application is being utilized or configured in your specific environment.

Having said that, if you think that's not the case and are interested in contributing a solution, we welcome you to create a pull request. The Bitnami team is excited to review your submission and offer feedback. You can find the contributing guidelines [here](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md).

Your contribution will greatly benefit the community. Feel free to reach out if you have any questions or need assistance.

If you have any questions about the application itself, customizing its content, or questions about technology and infrastructure usage, we highly recommend that you refer to the forums and user guides provided by the project responsible for the application or technology.

With that said, we'll keep this ticket open until the stale bot automatically closes it, in case someone from the community contributes valuable insights.

**Comment by LucaCourtial** [comment]
Hi, same problem with me, when I run "docker stop redis-master" here are the logs on my sentinels :

```
2023-12-01T08:47:55.033215032Z 1:X 01 Dec 2023 08:47:55.032 # Failed to resolve hostname 'redis'
2023-12-01T08:47:55.100261206Z 1:X 01 Dec 2023 08:47:55.099 # +tilt #tilt mode entered
2023-12-01T08:48:00.257121883Z 1:X 01 Dec 2023 08:48:00.256 # Failed to resolve hostname 'redis'
2023-12-01T08:48:00.316490340Z 1:X 01 Dec 2023 08:48:00.316 # +tilt #tilt mode entered
2023-12-01T08:48:03.366620690Z 1:X 01 Dec 2023 08:48:03.366 # Failed to resolve hostname 'redis'
2023-12-01T08:48:03.434211109Z 1:X 01 Dec 2023 08:48:03.433 # +tilt #tilt mode entered
2023-12-01T08:48:08.616366931Z 1:X 01 Dec 2023 08:48:08.616 # Failed to resolve hostname 'redis'
2023-12-01T08:48:08.683328099Z 1:X 01 Dec 2023 08:48:08.682 # +tilt #tilt mode entered
```

My slave :
```
2023-12-01T08:48:59.816808730Z 1:S 01 Dec 2023 08:48:59.816 * Connecting to MASTER redis:6379
2023-12-01T08:49:01.908510024Z 1:S 01 Dec 2023 08:49:01.908 # Unable to connect to MASTER: Invalid argument
2023-12-01T08:49:02.915821675Z 1:S 01 Dec 2023 08:49:02.915 * Connecting to MASTER redis:6379
2023-12-01T08:49:05.024477841Z 1:S 01 Dec 2023 08:49:05.024 # Unable to connect to MASTER: Invalid argument
2023-12-01T08:49:06.029186400Z 1:S 01 Dec 2023 08:49:06.028 * Connecting to MASTER redis:6379
2023-12-01T08:49:08.127813975Z 1:S 01 Dec 2023 08:49:08.126 # Unable to connect to MASTER: Invalid argument
```

However, I had this issue with redis image too, not only with bitnami

**Comment by oliviarla** [comment]
I don't know exactly how this problem solved, but in my case failover succeed when I set `REDIS_MASTER_HOST` to device's ip.

```yml
  redis-sentinel:
    image: 'bitnami/redis-sentinel:latest'
    environment:
      - REDIS_SENTINEL_DOWN_AFTER_MILLISECONDS=3000
      - REDIS_MASTER_HOST=192.168.1.110 # if this variable set to redis, failover not working.
      - REDIS_MASTER_PORT_NUMBER=6379
```

**Comment by gongomgra** [comment]
Hi @Luk7c  and @oliviarla

Thanks for posting your solution. As it is mentioned in a previous message that this issue is reproducible with official Redis image, please open a new ticket in the official Redis forums for them to provide further information on this. We will keep this thread open (I'm adding the "on hold" label).

**Comment by oliviarla** [comment]
@gongomgra 
Thank you for response. But I think the problem is redis-sentinel image from bitnami cannot find redis(master/replica) container's bridge network ip. (by `REDIS_MASTER_HOST`)
Because redis image from bitnami can lookup redis container's bridge network ip. (by `REDIS_MASTER_HOST`)

This is the docker-compose.yml.
```yml
networks:
  redis_network:
    driver: bridge

services:
  redis:
    image: 'bitnami/redis:latest'
    environment:
      - REDIS_REPLICATION_MODE=master
      - ALLOW_EMPTY_PASSWORD=yes
    ports:
      - 6379:6379
    networks:
      - redis_network
  redis-slave-1:
    image: 'bitnami/redis:latest'
    environment:
      - REDIS_REPLICATION_MODE=slave
      - REDIS_MASTER_HOST=redis
      - ALLOW_EMPTY_PASSWORD=yes
    ports:
      - 6479:6479
    depends_on:
      - redis
    networks:
      - redis_network
  redis-slave-2:
    image: 'bitnami/redis:latest'
    environment:
      - REDIS_REPLICATION_MODE=slave
      - REDIS_MASTER_HOST=redis
      - ALLOW_EMPTY_PASSWORD=yes
    ports:
      - 6579:6579
    depends_on:
      - redis
    networks:
      - redis_network
  redis-sentinel:
    image: 'bitnami/redis-sentinel:latest'
    environment:
      - REDIS_SENTINEL_DOWN_AFTER_MILLISECONDS=3000
      - REDIS_MASTER_HOST=192.168.1.110
      - REDIS_MASTER_PORT_NUMBER=6379
      - REDIS_MASTER_SET=mymaster
      - REDIS_SENTINEL_QUORUM=2
      - REDIS_SENTINEL_RESOLVE_HOSTNAMES=yes
    depends_on:
      - redis
      - redis-slave-1
      - redis-slave-2
    ports:
      - '26379-26381:26379'
    networks:
      - redis_network
```

**Comment by gongomgra** [comment]
Hi @oliviarla,

I think I have found the solution to your issue. The `REDIS_SENTINEL_QUORUM ` variable is set to 2, which means two sentinels instances are required to elect a new master in a failover situation. I tried two different scenarios: set quorum to `1`, and add a new sentinel service keeping quorum to `2`. In both cases, the cluster elected a new master out of the other redis nodes available without issues (I simply forced the failover workflow with the `sentinel failover mymaster` command to speed up the process after killing the original redis-primary node).

These are the logs electing a new primary node in my cluster

```
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:52.386 # +sdown master mymaster 172.27.0.2 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:52.386 # +odown master mymaster 172.27.0.2 6379 #quorum 1/1
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:52.386 # +new-epoch 1
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:52.386 # +try-failover master mymaster 172.27.0.2 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:52.389 * Sentinel new configuration saved on disk
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:52.389 # +vote-for-leader ea2bfc320b8baf2da0b47a9556d9c5eacb56c13a 1
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:52.389 # +elected-leader master mymaster 172.27.0.2 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:52.389 # +failover-state-select-slave master mymaster 172.27.0.2 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:52.460 # +selected-slave slave 172.27.0.3:6379 172.27.0.3 6379 @ mymaster 172.27.0.2 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:52.460 * +failover-state-send-slaveof-noone slave 172.27.0.3:6379 172.27.0.3 6379 @ mymaster 172.27.0.2 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:52.560 * +failover-state-wait-promotion slave 172.27.0.3:6379 172.27.0.3 6379 @ mymaster 172.27.0.2 6379
(...)
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:54.459 * +slave-reconf-inprog slave 172.27.0.4:6379 172.27.0.4 6379 @ mymaster 172.27.0.2 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:54.459 * +slave-reconf-done slave 172.27.0.4:6379 172.27.0.4 6379 @ mymaster 172.27.0.2 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:54.525 # +failover-end master mymaster 172.27.0.2 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:54.525 # +switch-master mymaster 172.27.0.2 6379 172.27.0.3 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:54.525 * +slave slave 172.27.0.4:6379 172.27.0.4 6379 @ mymaster 172.27.0.3 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:54.532 # Failed to resolve hostname 'redis'
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:54.532 * +slave slave :6379  6379 @ mymaster 172.27.0.3 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:54.535 * Sentinel new configuration saved on disk
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:54.593 # Failed to resolve hostname 'redis'
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:55.628 # Failed to resolve hostname 'redis'
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:56.632 # Failed to resolve hostname 'redis'
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:57.550 # +sdown slave :6379  6379 @ mymaster 172.27.0.3 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:57.656 # Failed to resolve hostname 'redis'
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:58.701 # Failed to resolve hostname 'redis'
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:28:59.753 # Failed to resolve hostname 'redis'
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:00.784 # Failed to resolve hostname 'redis'
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:01.867 # Failed to resolve hostname 'redis'
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:02.507 * Executing user requested FAILOVER of 'mymaster'
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:02.507 # +new-epoch 2
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:02.507 # +try-failover master mymaster 172.27.0.3 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:02.543 * Sentinel new configuration saved on disk
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:02.543 # +vote-for-leader ea2bfc320b8baf2da0b47a9556d9c5eacb56c13a 2
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:02.543 # +elected-leader master mymaster 172.27.0.3 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:02.543 # +failover-state-select-slave master mymaster 172.27.0.3 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:02.634 # +selected-slave slave 172.27.0.4:6379 172.27.0.4 6379 @ mymaster 172.27.0.3 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:02.634 * +failover-state-send-slaveof-noone slave 172.27.0.4:6379 172.27.0.4 6379 @ mymaster 172.27.0.3 6379
support-redis-sentinel-1-1  | 1:X 09 Jan 2024 11:29:02.689 * +failover-state-wait-promotion slave 172.27.0.4:6379 172.27.0.4 6379 @ mymaster 172.27.0.3 6379
```

And I also see the change in the sentinel cli

```
127.0.0.1:26379> sentinel master mymaster
 1) "name"
 2) "mymaster"
 3) "ip"
 4) "172.27.0.3"
 5) "port"
 6) "6379"
(...)
127.0.0.1:26379> SENTINEL FAILOVER mymaster
OK
127.0.0.1:26379> sentinel master mymaster
 1) "name"
 2) "mymaster"
 3) "ip"
 4) "172.27.0.4"
 5) "port"
 6) "6379"
```

Regarding the failed to resolve hostname error message, I understand that's something more related to Redis and Sentinel itself than to our configuration (as mentioned in other messages in this thread), but I hope the quorum tip helps you.

**Comment by ellyoana** [comment]
Hey all,

After reading this thread, I am still not sure what is the solution to the problem.

For context, I used a very similar setup to the one provided by @oliviarla and I also set the quorum size to 1 (even though I have 1 master, 2 slaves and 3 sentinels). Then, I shut down the master, followed by running `sentinel failover mymaster`. For the first few seconds, I got the same output as @gongomgra presented, then I kept getting `Failed to resolve hostname 'redis'` from all 3 sentinels.

My question is: is this the expected behavior, or could there be a misconfiguration?
